File: ios/TacticalMaps/Models/TrackRecorder.swift
import Foundation
import CoreLocation

/// One recorded fix on a GPX track.
struct TrackPoint {
    let coordinate: CLLocationCoordinate2D
    let elevation: Double?
    let time: Date
}

/// Truthful UI state for the user-started recording workflow. Permission
/// prompts and durable log preparation are intentionally distinct from an
/// active recording, so neither can accidentally display `REC`.
final class RecordingCoordinator: ObservableObject {
    enum State: Equatable {
        case idle
        case awaitingPermission
        case starting
        case recording
        case interrupted(String)
    }

    struct Guidance: Identifiable, Equatable {
        let id = UUID()
        let message: String
        let offersSettings: Bool

        static func == (lhs: Guidance, rhs: Guidance) -> Bool {
            lhs.message == rhs.message && lhs.offersSettings == rhs.offersSettings
        }
    }

    @Published private(set) var state: State = .idle
    @Published var guidance: Guidance?

    private let requestAuthorization: () -> Void
    private let initializeDurableRecording: () -> Bool
    private let stopRecording: () -> Void
    private let setBackgroundUpdates: (Bool) -> Void
    private let recordingError: () -> String?

    init(
        requestAuthorization: @escaping () -> Void,
        initializeDurableRecording: @escaping () -> Bool,
        stopRecording: @escaping () -> Void,
        setBackgroundUpdates: @escaping (Bool) -> Void,
        recordingError: @escaping () -> String?
    ) {
        self.requestAuthorization = requestAuthorization
        self.initializeDurableRecording = initializeDurableRecording
        self.stopRecording = stopRecording
        self.setBackgroundUpdates = setBackgroundUpdates
        self.recordingError = recordingError
    }

    func toggle(authorization: CLAuthorizationStatus) {
        switch state {
        case .recording, .awaitingPermission:
            stop()
        case .starting:
            break
        case .idle, .interrupted:
            start(authorization: authorization)
        }
    }

    func start(authorization: CLAuthorizationStatus) {
        guard state != .recording, state != .starting else { return }
        guidance = nil
        switch authorization {
        case .notDetermined:
            guard state != .awaitingPermission else { return }
            state = .awaitingPermission
            requestAuthorization()
        case .authorizedAlways, .authorizedWhenInUse:
            beginDurableStart()
        case .denied, .restricted:
            state = .idle
            guidance = Self.permissionGuidance(interrupted: false)
        @unknown default:
            state = .idle
            guidance = Self.permissionGuidance(interrupted: false)
        }
    }

    func authorizationChanged(_ authorization: CLAuthorizationStatus) {
        switch authorization {
        case .authorizedAlways, .authorizedWhenInUse:
            if state == .awaitingPermission { beginDurableStart() }
        case .denied, .restricted:
            if state == .recording || state == .starting {
                interruptForPermissionLoss()
            } else if state == .awaitingPermission {
                state = .idle
                guidance = Self.permissionGuidance(interrupted: false)
            }
        case .notDetermined:
            if state == .recording || state == .starting {
                interruptForPermissionLoss()
            }
        @unknown default:
            if state == .recording || state == .starting {
                interruptForPermissionLoss()
            }
        }
    }

    func recorderDidStopUnexpectedly(error: String?) {
        guard state == .recording else { return }
        setBackgroundUpdates(false)
        let message = error ?? L10n.text("Track recording was interrupted. The saved track was preserved.")
        state = .interrupted(message)
    }

    func stop() {
        let wasActive = state == .recording || state == .starting
        if wasActive { stopRecording() }
        setBackgroundUpdates(false)
        state = .idle
        guidance = nil
    }

    private func beginDurableStart() {
        guard state != .recording, state != .starting else { return }
        state = .starting
        guard initializeDurableRecording() else {
            setBackgroundUpdates(false)
            state = .interrupted(
                recordingError() ?? L10n.text("Track recording could not start. No recording is active.")
            )
            return
        }
        // TrackRecorder has created and fsync'd the log and captured its scoped
        // key at this point. Only now enable the conditional background mode
        // and publish the state from which the UI renders `REC`.
        setBackgroundUpdates(true)
        state = .recording
    }

    private func interruptForPermissionLoss() {
        stopRecording()
        setBackgroundUpdates(false)
        let message = L10n.text("Location access changed, so track recording stopped. Your saved track was preserved. Re-enable Location access in Settings to record again.")
        state = .interrupted(message)
        guidance = Guidance(message: message, offersSettings: true)
    }

    private static func permissionGuidance(interrupted: Bool) -> Guidance {
        Guidance(
            message: interrupted
                ? L10n.text("Track recording stopped because Location access is unavailable. Your saved track was preserved. Re-enable it in Settings.")
                : L10n.text("TacMap needs Location access to record a track. Allow access in Settings, then try again."),
            offersSettings: true
        )
    }
}

/// Accumulates GPS fixes into a track while recording. Fed by ContentView from
/// `LocationService.lastLocation`. Foreground + background (background only
/// while recording, via `LocationService.setBackgroundUpdates`).
///
/// Each fix gets appended + fsync'd to
/// `Application Support/tracks/recording.ndjson` as it arrives, so if the
/// process dies (OOM kill, crash, reboot) we lose at most one in-flight fix,
/// not the whole track. On init we check for a leftover log from a previous
/// session that wasnt cleanly discarded and recover it.
final class TrackRecorder: ObservableObject {
    @Published private(set) var isRecording = false
    @Published private(set) var points: [TrackPoint] = []
    /// True when `points` were recovered from a previous, un-discarded session.
    @Published private(set) var recovered = false

    /// Non-nil when a fix couldn't be written to disk. The UI has to say so: a
    /// recording that looks live but isn't hitting the disk is the worst
    /// possible failure for a field tool.
    @Published var persistError: String?
    @Published private(set) var requiresUnlock = false

    /// Bound in as AEAD associated data. Matches Android's TrackLog.LABEL.
    private static let label = "tracks/recording.ndjson"

    /// Min spacing between stored fixes. Drops GPS jitter so a stationary
    /// device doesn't bloat the track.
    private let minSpacingMetres: Double = 2
    private let maxRecoveryBytes = 64 * 1024 * 1024
    /// Scoped copy used only by a recording explicitly started by the user.
    /// The app may clear DataKey's global cache on background/App Lock without
    /// interrupting the already-authorised background recording.
    private var recordingKey: Data?

#if DEBUG
    /// Boolean-only test seam for scoped-key lifetime. Key material never
    /// crosses this boundary.
    var hasScopedRecordingKeyForTesting: Bool { recordingKey != nil }
#endif

    private let fileURL: URL
    private let attributeReader: (String) throws -> [FileAttributeKey: Any]
    private let textReader: (URL) throws -> String
    private let requiresSealedPolicy: (String, Data) throws -> Bool
    private let markSealedPolicy: (String, Data) throws -> Void

    private static func defaultFileURL() -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory,
                                            in: .userDomainMask).first!
        let dir = base.appendingPathComponent("tracks")
        return dir.appendingPathComponent("recording.ndjson")
    }

    /// On-disk line format. CLLocationCoordinate2D isn't Codable so we roll our own.
    private struct StoredPoint: Codable {
        let lat: Double
        let lon: Double
        let ele: Double?
        let t: Double // timeIntervalSince1970
    }

    init(
        fileURL: URL? = nil,
        attributeReader: @escaping (String) throws -> [FileAttributeKey: Any] = {
            try FileManager.default.attributesOfItem(atPath: $0)
        },
        textReader: @escaping (URL) throws -> String = {
            try String(contentsOf: $0, encoding: .utf8)
        },
        requiresSealedPolicy: @escaping (String, Data) throws -> Bool = {
            try SealedMigrationPolicy.requiresSealed($0, key: $1)
        },
        markSealedPolicy: @escaping (String, Data) throws -> Void = {
            try SealedMigrationPolicy.markSealed($0, key: $1)
        }
    ) {
        self.fileURL = fileURL ?? Self.defaultFileURL()
        self.attributeReader = attributeReader
        self.textReader = textReader
        self.requiresSealedPolicy = requiresSealedPolicy
        self.markSealedPolicy = markSealedPolicy
        recover()
    }

    @discardableResult
    func start() -> Bool {
        do {
            let key = try SafeStore.keyProvider()
            let dir = fileURL.deletingLastPathComponent()
            try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            do {
                let existingAttributes = try attributeReader(fileURL.path)
                guard let size = existingAttributes[.size] as? NSNumber else {
                    throw CocoaError(.fileReadUnknown)
                }
                guard size.int64Value == 0 else {
                    persistError = L10n.text("A saved track already exists. Export or discard it before starting a new recording.")
                    return false
                }
            } catch {
                // Explicit absence is the only state in which creating a new
                // log is safe. Permission/protection/metadata errors must not
                // silently become size zero and overwrite unknown bytes.
                guard Self.isNoSuchFile(error) else { throw error }
            }
            try markSealedPolicy(policyID, key)
            // Establish and fsync the durable log before telling the UI or
            // CLLocationManager that recording has begun.
            try Data().write(to: fileURL,
                             options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
            let handle = try FileHandle(forWritingTo: fileURL)
            try handle.synchronize()
            try handle.close()
            recordingKey = key
            points = []
            recovered = false
            persistError = nil
            requiresUnlock = false
            isRecording = true
            return true
        } catch {
            recordingKey = nil
            isRecording = false
            persistError = L10n.text("Recording did not start: %1$@", error.localizedDescription)
            return false
        }
    }

    func stop() {
        isRecording = false
        recordingKey = nil
        // Keep the log around so a finished-but-unexported track survives
        // process death until user exports or discards.
    }

    /// Clear the current (recorded or recovered) track and remove its file.
    func discard() {
        points = []
        recovered = false
        isRecording = false
        recordingKey = nil
        try? FileManager.default.removeItem(at: fileURL)
        try? FileManager.default.removeItem(at: legacyMarkerURL)
    }

    func ingest(_ location: CLLocation) {
        guard isRecording else { return }
        if let last = points.last {
            let prev = CLLocation(latitude: last.coordinate.latitude,
                                  longitude: last.coordinate.longitude)
            if prev.distance(from: location) < minSpacingMetres { return }
        }
        guard location.coordinate.latitude.isFinite,
              location.coordinate.longitude.isFinite,
              abs(location.coordinate.latitude) <= 90,
              abs(location.coordinate.longitude) <= 180 else { return }
        let point = TrackPoint(
            coordinate: location.coordinate,
            // verticalAccuracy < 0 means altitude is invalid.
            elevation: location.verticalAccuracy >= 0 ? location.altitude : nil,
            time: location.timestamp
        )
        do {
            try append(point) // persist before a shown fix becomes observable
            points.append(point)
            persistError = nil
        } catch {
            persistError = L10n.text("Track recording stopped because a fix could not be saved.")
            isRecording = false
            recordingKey = nil
        }
    }

    // MARK: - Persistence

    /// Each line is sealed on its own rather than sealing the whole file, for
    /// two reasons. Appending to a whole-file envelope would mean decrypt-all,
    /// re-encrypt-all, rewrite-all on every single GPS fix, which is O(n^2) over
    /// a patrol. And per-line sealing keeps the property we actually care about:
    /// a torn or garbled line fails its tag check and gets skipped, and every
    /// other line still opens. One bad byte costs you one fix, not the track.
    ///
    /// So we don't bind line ordering into the AAD. Someone with write access to
    /// the app sandbox could reorder or drop lines undetected. Took that trade
    /// knowingly: binding the index would make one torn line invalidate the whole
    /// tail, and an attacker inside our sandbox has already beaten at-rest crypto.
    private func append(_ point: TrackPoint) throws {
        let stored = StoredPoint(lat: point.coordinate.latitude,
                                 lon: point.coordinate.longitude,
                                 ele: point.elevation,
                                 t: point.time.timeIntervalSince1970)
        guard let recordingKey else { throw DataKey.LockedError() }
        let json = try JSONEncoder().encode(stored)
        let sealed = try SealedEnvelope.sealLine(key: recordingKey,
                                                 plaintext: json,
                                                 label: Self.label)
        guard let line = (sealed + "\n").data(using: .utf8) else {
            throw CocoaError(.fileWriteInapplicableStringEncoding)
        }
        if !FileManager.default.fileExists(atPath: fileURL.path) {
            guard FileManager.default.createFile(atPath: fileURL.path, contents: nil,
                attributes: [.protectionKey: FileProtectionType.completeUntilFirstUserAuthentication]) else {
                throw CocoaError(.fileWriteUnknown)
            }
        }
        let handle = try FileHandle(forWritingTo: fileURL)
        defer { try? handle.close() }
        handle.seekToEndOfFile()
        try handle.write(contentsOf: line)
        try handle.synchronize() // fsync, force to stable storage
    }

    private func recover() {
        let attrs: [FileAttributeKey: Any]
        do {
            attrs = try attributeReader(fileURL.path)
        } catch {
            if Self.isNoSuchFile(error) {
                requiresUnlock = false
                return
            }
            markRecoveryUnavailable(error)
            return
        }
        guard let size = attrs[.size] as? NSNumber, size.int64Value >= 0 else {
            markRecoveryUnavailable(CocoaError(.fileReadUnknown))
            return
        }
        guard size.int64Value <= Int64(maxRecoveryBytes) else {
            persistError = L10n.text("Saved track is too large to recover safely.")
            return
        }
        let text: String
        do {
            text = try textReader(fileURL)
        } catch {
            if Self.isNoSuchFile(error) {
                requiresUnlock = false
                return
            }
            markRecoveryUnavailable(error)
            return
        }
        let key: Data
        do {
            key = try SafeStore.keyProvider()
        } catch {
            // Locked. The log is intact, we just can't read it yet. Say nothing
            // about "no track" because we genuinely don't know.
            persistError = L10n.text("Saved track is encrypted and locked. %1$@", error.localizedDescription)
            requiresUnlock = true
            return
        }
        requiresUnlock = false

        let decoder = JSONDecoder()
        let legacyMarker = FileManager.default.fileExists(atPath: legacyMarkerURL.path)
        if legacyMarker {
            do {
                try markSealedPolicy(policyID, key)
            } catch {
                markRecoveryUnavailable(error)
                return
            }
        }
        let sealedOnly: Bool
        do {
            sealedOnly = try requiresSealedPolicy(policyID, key)
        } catch {
            markRecoveryUnavailable(error)
            return
        }
        if sealedOnly && text.split(separator: "\n").contains(where: {
            !SealedEnvelope.isSealedLine(String($0))
        }) {
            persistError = L10n.text("Saved track failed its sealed-only integrity check.")
            return
        }
        var sawLegacyLine = false
        var invalidLegacyLine = false
        let restored: [TrackPoint] = text.split(separator: "\n").compactMap { raw in
            let line = String(raw)
            let json: Data?
            if SealedEnvelope.isSealedLine(line) {
                json = SealedEnvelope.openLine(key: key, line: line, label: Self.label)
            } else {
                sawLegacyLine = true // written by a pre-encryption build
                json = line.data(using: .utf8)
            }
            guard let json, let sp = try? decoder.decode(StoredPoint.self, from: json) else {
                if !SealedEnvelope.isSealedLine(line) { invalidLegacyLine = true }
                else { persistError = L10n.text("Saved track contains an authenticated line that could not be recovered.") }
                return nil
            }
            guard sp.lat.isFinite, sp.lon.isFinite, sp.t.isFinite,
                  abs(sp.lat) <= 90, abs(sp.lon) <= 180,
                  sp.ele?.isFinite != false else {
                if !SealedEnvelope.isSealedLine(line) { invalidLegacyLine = true }
                return nil
            }
            return TrackPoint(coordinate: CLLocationCoordinate2D(latitude: sp.lat, longitude: sp.lon),
                              elevation: sp.ele,
                              time: Date(timeIntervalSince1970: sp.t))
        }
        if sawLegacyLine {
            if invalidLegacyLine {
                persistError = L10n.text("Saved legacy track contains an invalid line and was preserved unchanged.")
                return
            } else {
                guard reseal(restored, key: key) else { return }
            }
        } else if !text.isEmpty {
            do {
                // Existing ciphertext from a build predating the authenticated
                // policy is not observable until the durable downgrade fence
                // exists. A policy-store failure leaves the log untouched and
                // retryable instead of silently accepting later plaintext.
                try markSealedPolicy(policyID, key)
            } catch {
                markRecoveryUnavailable(error)
                return
            }
            if legacyMarker { try? FileManager.default.removeItem(at: legacyMarkerURL) }
        }
        if !restored.isEmpty {
            points = restored
            recovered = true
        }
    }

    /// Rewrite a plaintext log from an older build with every line sealed. Goes
    /// via atomic write so a crash halfway can't leave us half a track.
    @discardableResult
    private func reseal(_ points: [TrackPoint], key: Data) -> Bool {
        do {
            let encoder = JSONEncoder()
            var out = Data()
            for point in points {
                let stored = StoredPoint(lat: point.coordinate.latitude,
                                         lon: point.coordinate.longitude,
                                         ele: point.elevation,
                                         t: point.time.timeIntervalSince1970)
                let sealed = try SealedEnvelope.sealLine(key: key,
                                                         plaintext: try encoder.encode(stored),
                                                         label: Self.label)
                out.append(contentsOf: Array((sealed + "\n").utf8))
            }
            // Fence first. If the subsequent atomic replacement fails, the
            // intact plaintext is deliberately no longer accepted on relaunch.
            try markSealedPolicy(policyID, key)
            try out.write(to: fileURL, options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
            if FileManager.default.fileExists(atPath: legacyMarkerURL.path) {
                try? FileManager.default.removeItem(at: legacyMarkerURL)
            }
            return true
        } catch {
            requiresUnlock = true
            persistError = L10n.text("Could not encrypt the recovered track: %1$@", error.localizedDescription)
            return false
        }
    }

    func retryRecoveryAfterUnlock() {
        guard requiresUnlock else { return }
        persistError = nil
        recover()
    }

    private func markRecoveryUnavailable(_ error: Error) {
        requiresUnlock = true
        persistError = L10n.text("Saved track is protected or unavailable and was left untouched. Retry after unlocking the device. %1$@", error.localizedDescription)
    }

    private static func isNoSuchFile(_ error: Error) -> Bool {
        var pending = [error as NSError]
        var visited = Set<ObjectIdentifier>()

        while let nsError = pending.popLast() {
            guard visited.insert(ObjectIdentifier(nsError)).inserted else { continue }

            if nsError.domain == NSCocoaErrorDomain,
               (nsError.code == NSFileNoSuchFileError
                || nsError.code == CocoaError.Code.fileReadNoSuchFile.rawValue) {
                return true
            }
            if nsError.domain == NSPOSIXErrorDomain,
               nsError.code == POSIXError.Code.ENOENT.rawValue {
                return true
            }
            if let underlying = nsError.userInfo[NSUnderlyingErrorKey] as? NSError {
                pending.append(underlying)
            }
        }
        return false
    }

    private var legacyMarkerURL: URL { fileURL.appendingPathExtension("sealed-only") }
    private var policyID: String { "track:\(fileURL.standardizedFileURL.path):\(Self.label)" }
}

File: android/app/src/main/java/com/tacmap/models/TrackRecorder.kt
package com.tacmap.models

import android.content.Context
import android.location.Location
import android.os.Handler
import android.os.Looper
import com.tacmap.util.DataKey
import com.tacmap.util.SafeStore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import java.io.File
import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/** One recorded fix on a GPX track. */
@Serializable
data class TrackPoint(
    val latitude: Double,
    val longitude: Double,
    val elevationMetres: Double?,
    val timeEpochMs: Long
)

/**
 * Accumulates GPS fixes into a track. Fed from MapViewModel via
 * LocationService.
 *
 * Every fix is fsync'd to tracks/recording.ndjson via TrackLog so
 * process death only loses the single in-flight fix, never the whole
 * track. On construction we recover any leftover track from a session
 * that died mid-recording. Background recording kept alive by
 * TrackRecordingService.
 */
class TrackRecorder internal constructor(
    private val logFile: File,
    private val stopRecordingService: () -> Unit = {},
    private val scheduleActivationTimeout: (Long, () -> Unit) -> Unit = { _, _ -> },
    private val recordingKeyProvider: () -> ByteArray = { SafeStore.keyProvider.key().copyOf() },
) {

    constructor(context: Context) : this(
        logFile = File(context.applicationContext.filesDir, "tracks/recording.ndjson"),
        recordingKeyProvider = DataKey::key,
        stopRecordingService = { TrackRecordingService.stop(context.applicationContext) },
        scheduleActivationTimeout = { delayMillis, action ->
            Handler(Looper.getMainLooper()).postDelayed(action, delayMillis)
        },
    )

    private val recordingKeyLock = Any()
    private var recordingKey: ByteArray? = null
    private var nextSessionGeneration = 0L
    private var currentSessionGeneration: Long? = null

    private val _uiState = MutableStateFlow(TrackRecordingUiState())
    val uiState: StateFlow<TrackRecordingUiState> = _uiState.asStateFlow()

    private val _isRecording = MutableStateFlow(false)
    val isRecording: StateFlow<Boolean> = _isRecording.asStateFlow()

    private val _points = MutableStateFlow<List<TrackPoint>>(emptyList())
    val points: StateFlow<List<TrackPoint>> = _points.asStateFlow()

    /** True if we recovered points from a previous session that got killed
     *  mid-recording. UI can offer export/clear. */
    private val _recovered = MutableStateFlow(false)
    val recovered: StateFlow<Boolean> = _recovered.asStateFlow()

    /** Non-null when a fix couldn't be written to disk. The UI has to say so:
     *  a recording that looks live but isn't hitting the disk is the worst
     *  possible failure for a field tool. */
    private val _persistError = MutableStateFlow<String?>(null)
    val persistError: StateFlow<String?> = _persistError.asStateFlow()

    /** False until the existing log has been read and any legacy migration has
     *  durably completed. A failed verification must never be truncated by a
     *  subsequent start attempt. */
    private var recoveryReady = false

    fun acknowledgePersistError() { _persistError.value = null }

    fun dismissRecordingMessage() {
        _persistError.value = null
        if (_uiState.value.phase == TrackRecordingPhase.AwaitingPermission ||
            _uiState.value.phase == TrackRecordingPhase.Interrupted
        ) {
            _uiState.value = TrackRecordingReducer.reduce(
                _uiState.value,
                TrackRecordingEvent.Stopped,
            )
        }
    }

    /** Min spacing between fixes (m). Drops GPS jitter. */
    private val minSpacingMetres = 2.0

    init { recoverAfterUnlock() }

    /** Retry recovery after the platform credential has unwrapped the DEK. */
    fun reloadAfterUnlock() {
        if (_isRecording.value) return
        recoverAfterUnlock()
    }

    private fun recoverAfterUnlock() {
        // A locked at-rest key means we can't read the log yet. Leave it alone,
        // don't report an empty track as if the recording was lost.
        runCatching { TrackLog.read(logFile) }
            .onSuccess { restored ->
                if (restored.points.isNotEmpty()) {
                    _points.value = restored.points
                    _recovered.value = true
                }
                // Log came from a pre-encryption build, seal it in place once.
                if (restored.hadLegacyLines) {
                    runCatching { TrackLog.reseal(logFile, restored.points) }
                        .onSuccess { recoveryReady = true }
                        .onFailure {
                            recoveryReady = false
                            _persistError.value = "Could not encrypt the recovered track: ${it.message}"
                        }
                } else {
                    recoveryReady = true
                }
            }
            .onFailure {
                recoveryReady = false
                _persistError.value = "Could not read the saved track: ${it.message}"
            }
    }

    /** Resolve permission/GPS prerequisites before any file or key work begins. */
    fun awaitPermissionRequest(message: String) {
        _uiState.value = TrackRecordingReducer.reduce(
            _uiState.value,
            TrackRecordingEvent.AwaitingPermission(message),
        )
    }

    fun requestStart(locationAccess: LocationAccess, gpsEnabled: Boolean): Boolean {
        if (_uiState.value.isAuthorizedSession) return false
        _uiState.value = TrackRecordingReducer.reduce(
            _uiState.value,
            TrackRecordingEvent.StartRequested(locationAccess, gpsEnabled),
        )
        return _uiState.value.phase == TrackRecordingPhase.Starting
    }

    /**
     * Create/fsync the new log first, then retain a private copy of the current
     * mission DEK for this explicitly-authorized recording session. The service
     * must still call [onServiceActivated] before the UI can show REC.
     */
    fun prepareStart(): Boolean {
        if (_uiState.value.phase != TrackRecordingPhase.Starting) return false
        if (hasRetainedRecordingKey()) return false
        if (!recoveryReady) {
            failRecording(
                _persistError.value ?: "Could not verify the saved track before recording."
            )
            return false
        }
        _persistError.value = null
        if (_recovered.value || _points.value.isNotEmpty()) {
            failRecording("Export or discard the saved track before starting a new recording.")
            return false
        }
        var preparedGeneration: Long? = null
        return runCatching { TrackLog.truncate(logFile) }
            .mapCatching {
                val supplied = recordingKeyProvider()
                try {
                    require(supplied.size == 32) { "Mission data key must be 256 bits" }
                    synchronized(recordingKeyLock) {
                        clearRecordingKeyLocked()
                        recordingKey = supplied.copyOf()
                        nextSessionGeneration += 1
                        currentSessionGeneration = nextSessionGeneration
                        preparedGeneration = currentSessionGeneration
                    }
                } finally {
                    supplied.fill(0)
                }
            }
            .fold(
                onSuccess = {
                    _points.value = emptyList()
                    _recovered.value = false
                    _isRecording.value = false
                    val generation = checkNotNull(preparedGeneration)
                    scheduleActivationTimeout(ACTIVATION_TIMEOUT_MILLIS) {
                        onServiceActivationTimedOut(generation)
                    }
                    true
                },
                onFailure = {
                    failRecording("Could not start recording safely: ${it.message}")
                    false
                }
            )
    }

    /** Called only after startForeground and the GPS listener both succeed. */
    fun onServiceActivated(generation: Long): Boolean {
        val matchesPreparedSession = synchronized(recordingKeyLock) {
            recordingKey != null && currentSessionGeneration == generation
        }
        if (_uiState.value.phase != TrackRecordingPhase.Starting || !matchesPreparedSession) return false
        _uiState.value = TrackRecordingReducer.reduce(
            _uiState.value,
            TrackRecordingEvent.ServiceActivated,
        )
        _isRecording.value = _uiState.value.phase == TrackRecordingPhase.Recording
        return _isRecording.value
    }

    internal fun preparedServiceGeneration(): Long? = synchronized(recordingKeyLock) {
        currentSessionGeneration.takeIf {
            _uiState.value.phase == TrackRecordingPhase.Starting && recordingKey != null
        }
    }

    internal fun isServiceSessionAuthorized(generation: Long): Boolean =
        synchronized(recordingKeyLock) {
            currentSessionGeneration == generation && recordingKey != null &&
                _uiState.value.isAuthorizedSession
        }

    internal fun onServiceActivationTimedOut(generation: Long) {
        if (_uiState.value.phase != TrackRecordingPhase.Starting ||
            !isServiceSessionAuthorized(generation)
        ) return
        interruptRecording(
            "Background recording did not activate in time. Try starting it again.",
            requestServiceStop = true,
        )
    }

    /** Called from Service.onDestroy; never asks the already-destroying service to stop again. */
    internal fun onServiceDestroyed(generation: Long) {
        if (!isServiceSessionAuthorized(generation)) return
        interruptRecording(
            "Background recording stopped unexpectedly. Your saved track was preserved.",
            requestServiceStop = false,
        )
    }

    internal fun onServiceFailure(
        generation: Long,
        message: String,
        settingsTarget: TrackRecordingSettingsTarget? = null,
    ) {
        if (!isServiceSessionAuthorized(generation)) return
        interruptRecording(message, settingsTarget, requestServiceStop = false)
    }

    fun stop() {
        _isRecording.value = false
        _uiState.value = TrackRecordingReducer.reduce(_uiState.value, TrackRecordingEvent.Stopped)
        clearRecordingKey()
        // Intentionally keep the log file - a completed track must survive
        // until user exports or discards it.
    }

    /** Clear a stopped (recorded or recovered) track and remove its file.
     *
     * Active recordings must first go through the explicit stop/confirmation
     * flow in the UI. Refusing here as well prevents a future caller from
     * silently deleting a live patrol track. State is only cleared after the
     * encrypted log has actually been removed. */
    fun discard(): Boolean {
        if (_isRecording.value) {
            _persistError.value = "Stop recording before discarding the saved track."
            return false
        }
        clearRecordingKey()
        return runCatching { TrackLog.delete(logFile) }
            .fold(
                onSuccess = {
                    _points.value = emptyList()
                    _recovered.value = false
                    _persistError.value = null
                    true
                },
                onFailure = {
                    _persistError.value = "Could not discard the saved track: ${it.message}"
                    false
                }
            )
    }

    fun onLocation(loc: Location) {
        recordPoint(
            TrackPoint(
                latitude = loc.latitude,
                longitude = loc.longitude,
                elevationMetres = if (loc.hasAltitude()) loc.altitude else null,
                timeEpochMs = if (loc.time > 0) loc.time else System.currentTimeMillis()
            )
        )
    }

    /** Shared persistence path for device fixes and host-side state-machine
     * tests. The point is not published until its encrypted append is durable. */
    internal fun recordPoint(point: TrackPoint) {
        if (!_isRecording.value) return
        if (!point.latitude.isFinite() || !point.longitude.isFinite() ||
            point.latitude !in -90.0..90.0 || point.longitude !in -180.0..180.0
        ) return
        val last = _points.value.lastOrNull()
        if (last != null &&
            distanceMetres(last.latitude, last.longitude, point.latitude, point.longitude) < minSpacingMetres
        ) return
        // Persist before publishing the fix. If durability fails, recording
        // stops instead of presenting an in-memory-only track as live.
        val failure = synchronized(recordingKeyLock) {
            val key = recordingKey
                ?: return@synchronized IllegalStateException("Recording key is unavailable")
            runCatching { TrackLog.append(logFile, point, key) }.exceptionOrNull()
        }
        if (failure == null) {
            _points.value = _points.value + point
        } else {
            failRecording("Track fix not saved; recording stopped: ${failure.message}")
        }
    }

    fun failRecording(
        message: String,
        settingsTarget: TrackRecordingSettingsTarget? = null,
    ) {
        interruptRecording(message, settingsTarget, requestServiceStop = true)
    }

    fun onLocationAccessChanged(locationAccess: LocationAccess) {
        val next = TrackRecordingReducer.reduce(
            _uiState.value,
            TrackRecordingEvent.PermissionChanged(locationAccess),
        )
        if (next == _uiState.value) return
        _isRecording.value = false
        _uiState.value = next
        _persistError.value = next.message
        clearRecordingKey()
        stopRecordingService()
    }

    /** Activity locking clears the general DEK, not an authorized session key. */
    fun onMissionKeyLock() {
        val authorized = synchronized(recordingKeyLock) {
            _uiState.value.isAuthorizedSession && recordingKey != null
        }
        if (authorized) return
        clearRecordingKey()
        if (_uiState.value.isAuthorizedSession) {
            failRecording("Recording stopped because its session key was unavailable.")
        }
    }

    internal fun hasRetainedRecordingKey(): Boolean =
        synchronized(recordingKeyLock) { recordingKey != null }

    private fun clearRecordingKey() = synchronized(recordingKeyLock) {
        clearRecordingKeyLocked()
    }

    private fun clearRecordingKeyLocked() {
        recordingKey?.fill(0)
        recordingKey = null
        currentSessionGeneration = null
    }

    private fun interruptRecording(
        message: String,
        settingsTarget: TrackRecordingSettingsTarget? = null,
        requestServiceStop: Boolean,
    ) {
        _isRecording.value = false
        _persistError.value = message
        _uiState.value = TrackRecordingReducer.reduce(
            _uiState.value,
            TrackRecordingEvent.Interrupted(message, settingsTarget),
        )
        clearRecordingKey()
        if (requestServiceStop) stopRecordingService()
    }

    private companion object {
        const val ACTIVATION_TIMEOUT_MILLIS = 10_000L
    }

    private fun distanceMetres(aLat: Double, aLng: Double, bLat: Double, bLng: Double): Double {
        val lat1 = Math.toRadians(aLat)
        val lat2 = Math.toRadians(bLat)
        val dLat = lat2 - lat1
        val dLng = Math.toRadians(bLng - aLng)
        val h = sin(dLat / 2) * sin(dLat / 2) +
            cos(lat1) * cos(lat2) * sin(dLng / 2) * sin(dLng / 2)
        return 2 * 6_371_000.0 * asin(sqrt(h.coerceIn(0.0, 1.0)))
    }
}
